File: server/src/main/java/net/tachyon/network/netty/codec/PacketDecoder.java
package net.tachyon.network.netty.codec;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.ByteToMessageDecoder;
import net.tachyon.network.netty.packet.InboundPacket;
import net.tachyon.utils.Utils;

import java.util.List;

public class PacketDecoder extends ByteToMessageDecoder {

    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf buf, List<Object> list) {
        if (buf.readableBytes() > 0) {
            final int packetId = Utils.readVarInt(buf);
            list.add(new InboundPacket(packetId, buf));
        }
    }
}

File: Network/Netty/Codec/PacketDecoder.cs
using DotNetty.Buffers;
using DotNetty.Codecs;
using DotNetty.Transport.Channels;
using Tachyon.Utils;

namespace Tachyon.Network.Netty.Codec;

public class PacketDecoder : ByteToMessageDecoder
{
    protected override void Decode(IChannelHandlerContext context, IByteBuffer input, List<object> output)
    {
        if (input.ReadableBytes < 1)
        {
            return;
        }

        int packetId = BufUtil.readVarInt(input);
        output.Add(new InboundPacket(packetId, input));
    }
}